File: backend/app/services/task_service.py
"""
Task service layer for business logic.
Handles task operations and enforces business rules.
"""
from typing import List, Optional, Dict, Any
from sqlalchemy.ext.asyncio import AsyncSession
from app.repositories.task_repository import TaskRepository
from app.models.database import Task, User, TaskStatus, TaskPriority
import uuid
from datetime import datetime

class TaskService:
    """Service layer for task business logic"""
    
    def __init__(self, session: AsyncSession):
        self.session = session
        self.repository = TaskRepository(session)
    
    async def create_task(
        self,
        title: str,
        user_id: uuid.UUID,
        project_id: uuid.UUID,
        description: Optional[str] = None,
        category_id: Optional[uuid.UUID] = None,
        assignee_id: Optional[uuid.UUID] = None,
        priority: TaskPriority = TaskPriority.MEDIUM,
        due_date: Optional[datetime] = None
    ) -> Task:
        """Create a new task with business logic validation"""
        
        # Business rule: If no assignee specified, assign to creator
        if not assignee_id:
            assignee_id = user_id
        
        # Create task entity
        task = Task(
            title=title.strip(),
            description=description,
            project_id=project_id,
            category_id=category_id,
            creator_id=user_id,
            assignee_id=assignee_id,
            status=TaskStatus.TODO,
            priority=priority,
            due_date=due_date,
            completed=False
        )
        
        return await self.repository.create(task)
    
    async def get_user_tasks(
        self,
        user_id: uuid.UUID,
        filters: Optional[Dict[str, Any]] = None
    ) -> List[Task]:
        """Get tasks for a user with optional filtering"""
        
        if not filters:
            filters = {}
        
        return await self.repository.get_user_tasks(
            user_id=user_id,
            limit=filters.get('limit', 50),
            offset=filters.get('offset', 0),
            status=filters.get('status'),
            priority=filters.get('priority'),
            category_id=filters.get('category_id')
        )
    
    async def complete_task(
        self,
        task_id: uuid.UUID,
        user_id: uuid.UUID
    ) -> bool:
        """Complete a single task with business logic"""
        
        # Verify user has access to the task
        task = await self.repository.get_by_id(task_id, user_id)
        if not task:
            return False
        
        # Business rule: Can only complete TODO or IN_PROGRESS tasks
        if task.status == TaskStatus.DONE:
            return False  # Already completed
        
        result = await self.repository.bulk_complete([task_id], user_id, True)
        return result > 0
    
    async def bulk_complete_tasks(
        self,
        task_ids: List[uuid.UUID],
        user_id: uuid.UUID,
        completed: bool = True
    ) -> Dict[str, Any]:
        """Bulk complete/uncomplete tasks with business validation"""
        
        # Validate all tasks exist and user has access
        accessible_tasks = []
        for task_id in task_ids:
            task = await self.repository.get_by_id(task_id, user_id)
            if task:
                accessible_tasks.append(task_id)
        
        if not accessible_tasks:
            return {
                "requested_count": len(task_ids),
                "affected_count": 0,
                "success": False,
                "message": "No accessible tasks found"
            }
        
        affected_count = await self.repository.bulk_complete(
            accessible_tasks, user_id, completed
        )
        
        return {
            "requested_count": len(task_ids),
            "affected_count": affected_count,
            "success": affected_count > 0,
            "message": f"{'Completed' if completed else 'Uncompleted'} {affected_count} tasks"
        }
    
    async def bulk_update_tasks(
        self,
        task_ids: List[uuid.UUID],
        user_id: uuid.UUID,
        update_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Bulk update tasks with business validation"""
        
        # Validate update data
        allowed_fields = {
            'title', 'description', 'status', 'priority', 
            'category_id', 'assignee_id', 'due_date'
        }
        
        filtered_data = {
            k: v for k, v in update_data.items() 
            if k in allowed_fields
        }
        
        if not filtered_data:
            return {
                "requested_count": len(task_ids),
                "affected_count": 0,
                "success": False,
                "message": "No valid fields to update"
            }
        
        affected_count = await self.repository.bulk_update(
            task_ids, filtered_data, user_id
        )
        
        return {
            "requested_count": len(task_ids),
            "affected_count": affected_count,
            "success": affected_count > 0,
            "message": f"Updated {affected_count} tasks"
        }
    
    async def get_task_statistics(self, user_id: uuid.UUID) -> Dict[str, Any]:
        """Get comprehensive task statistics"""
        
        stats = await self.repository.get_task_stats(user_id)
        
        # Add calculated fields
        total = stats.get('total', 0)
        completed = stats.get('completed', 0)
        
        completion_rate = (completed / total * 100) if total > 0 else 0
        
        return {
            **stats,
            "completion_rate": round(completion_rate, 1),
            "remaining": total - completed
        }
    
    async def assign_task(
        self,
        task_id: uuid.UUID,
        assignee_id: uuid.UUID,
        user_id: uuid.UUID
    ) -> bool:
        """Assign a task to a different user"""
        
        # Business rule: Only task creator can reassign tasks
        task = await self.repository.get_by_id(task_id, user_id)
        if not task or task.creator_id != user_id:
            return False
        
        result = await self.repository.bulk_update(
            [task_id],
            {"assignee_id": assignee_id},
            user_id
        )
        
        return result > 0

File: frontend/src/services/taskService.ts
import api from './api'
import { 
  type Task, 
  type CreateTaskRequest, 
  type UpdateTaskRequest, 
  type TasksResponse, 
  TaskPriority,
  TaskStatus
} from '../types/task'

// Re-export Task type for components that import from this service
export type { Task } from '../types/task'

export class TaskService {
  // Get all tasks with optional filtering
  static async getTasks(params?: {
    project_id?: string
    status?: TaskStatus
    assignee_id?: string
    page?: number
    per_page?: number
    search?: string
    offset?: number
  }): Promise<TasksResponse> {
    const response = await api.get('/tasks/', { params })
    return response.data
  }

  // Get a single task by ID
  static async getTask(id: string): Promise<Task> {
    const response = await api.get(`/tasks/${id}`)
    return response.data
  }

  // Alias for getTask (for consistency)
  static async getById(id: string): Promise<Task> {
    return this.getTask(id)
  }

  // Create a new task
  static async createTask(taskData: CreateTaskRequest): Promise<Task> {
    // Format dates to YYYY-MM-DD if they exist
    const formattedData = { ...taskData }
    if (formattedData.start_date) {
      formattedData.start_date = formattedData.start_date.split('T')[0]
    }
    if (formattedData.due_date) {
      formattedData.due_date = formattedData.due_date.split('T')[0]
    }
    const response = await api.post('/tasks/', formattedData)
    return response.data
  }

  // Update an existing task
  static async updateTask(id: string, taskData: UpdateTaskRequest): Promise<Task> {
    // Format dates to YYYY-MM-DD if they exist
    const formattedData = { ...taskData }
    if (formattedData.start_date) {
      formattedData.start_date = formattedData.start_date.split('T')[0]
    }
    if (formattedData.due_date) {
      formattedData.due_date = formattedData.due_date.split('T')[0]
    }
    const response = await api.put(`/tasks/${id}`, formattedData)
    return response.data
  }

  // Alias for updateTask (for consistency)
  static async update(id: string, taskData: UpdateTaskRequest): Promise<Task> {
    return this.updateTask(id, taskData)
  }

  // Delete a task
  static async deleteTask(id: string): Promise<void> {
    await api.delete(`/tasks/${id}`)
  }

  // Get task statistics
  static async getTaskStats(projectId?: string): Promise<{
    total: number
    completed: number
    in_progress: number
    todo: number
    overdue: number
    completion_rate: number
    urgent?: number
    high?: number
    medium?: number
    low?: number
    priority_breakdown?: Record<string, number>
  }> {
    try {
      const params = projectId ? { project_id: projectId } : {}
      const response = await api.get('/tasks/stats', { params })
      // Backend already returns the shape the Dashboard expects
      return response.data
    } catch (error: any) {
      // Return default stats if endpoint not available
      if (error?.response?.status === 404) {
        // Get tasks and calculate stats manually
        const tasksResponse = await this.getTasks({ project_id: projectId })
        const tasks = tasksResponse.tasks || []

        const stats = {
          total: tasks.length,
          todo: tasks.filter(t => t.status === 'todo').length,
          in_progress: tasks.filter(t => t.status === 'in_progress').length,
          completed: tasks.filter(t => t.status === 'completed').length,
          overdue: tasks.filter(t => {
            if (!t.due_date || t.status === 'completed') return false
            return new Date(t.due_date) < new Date()
          }).length,
          urgent: tasks.filter(t => t.priority === 'urgent').length,
          high: tasks.filter(t => t.priority === 'high').length,
          medium: tasks.filter(t => t.priority === 'medium').length,
          low: tasks.filter(t => t.priority === 'low').length,
          completion_rate: tasks.length > 0
            ? tasks.filter(t => t.status === 'completed').length / tasks.length
            : 0,
          priority_breakdown: {
            urgent: tasks.filter(t => t.priority === 'urgent').length,
            high: tasks.filter(t => t.priority === 'high').length,
            medium: tasks.filter(t => t.priority === 'medium').length,
            low: tasks.filter(t => t.priority === 'low').length
          }
        }

        return stats
      }
      throw error
    }
  }

  // Fetch all tasks by paging until has_more is false
  static async getAllTasks(params?: {
    project_id?: string
    status?: TaskStatus
    assignee_id?: string
    per_page?: number
    search?: string
  }): Promise<TasksResponse> {
    const perPage = Math.min(params?.per_page ?? 100, 100)
    let offset = 0
    const all: TasksResponse = { tasks: [], total: 0, limit: perPage, offset: 0, has_more: false }
    const seenTaskIds = new Set<string>()

    // Loop with safety cap
    for (let i = 0; i < 20; i++) {
      const page = await this.getTasks({ ...params, per_page: perPage, offset })

      // Filter out duplicates based on task ID
      const uniqueTasks = page.tasks.filter(task => {
        if (seenTaskIds.has(task.id)) {
          return false
        }
        seenTaskIds.add(task.id)
        return true
      })

      all.tasks = all.tasks.concat(uniqueTasks)
      all.total = page.total

      if (!page.has_more) {
        all.has_more = false
        break
      }
      offset += perPage
      all.has_more = true
    }
    return all
  }

  // Update task status (for drag and drop)
  static async updateTaskStatus(id: string, status: TaskStatus, position?: number): Promise<Task> {
    const updateData: UpdateTaskRequest = { status }
    if (position !== undefined) {
      updateData.position = position
    }
    return this.updateTask(id, updateData)
  }

  // Bulk update tasks
  static async bulkUpdateTasks(taskIds: string[], updates: UpdateTaskRequest): Promise<void> {
    await api.post('/bulk/update', {
      task_ids: taskIds,
      updates
    })
  }

  // Bulk complete tasks
  static async bulkCompleteTasks(taskIds: string[]): Promise<void> {
    await api.post('/bulk/complete', {
      task_ids: taskIds
    })
  }

  // Bulk change status
  static async bulkChangeStatus(taskIds: string[], status: TaskStatus): Promise<void> {
    await api.post('/bulk/status', {
      task_ids: taskIds,
      new_status: status
    })
  }

  // Search tasks with advanced filters
  static async searchTasks(params: {
    query?: string
    status?: TaskStatus[]
    priority?: TaskPriority[]
    assignee_id?: string
    category_id?: string
    project_id?: string
    start_date?: string
    due_date?: string
    completed?: boolean
    tags?: string[]
    sort_by?: string
    sort_order?: 'asc' | 'desc'
    limit?: number
    offset?: number
  }): Promise<TasksResponse> {
    const response = await api.get('/tasks/search', { params })
    return response.data
  }

  // Complete or uncomplete a task using dedicated endpoint
  static async completeTask(id: string, completed: boolean = true): Promise<Task> {
    const response = await api.post(`/tasks/${id}/complete`, null, {
      params: { completed }
    })
    return response.data
  }

  // Bulk delete tasks
  static async bulkDeleteTasks(taskIds: string[], hardDelete: boolean = false): Promise<void> {
    await api.post('/bulk/delete', {
      task_ids: taskIds,
      hard_delete: hardDelete
    })
  }

  // Bulk change priority
  static async bulkChangePriority(taskIds: string[], priority: TaskPriority): Promise<void> {
    await api.post('/bulk/priority', {
      task_ids: taskIds,
      new_priority: priority
    })
  }

  // Bulk assign category
  static async bulkAssignCategory(taskIds: string[], categoryId: string | null): Promise<void> {
    await api.post('/bulk/category', {
      task_ids: taskIds,
      category_id: categoryId
    })
  }
}